google/detectors/credentials/generic_weak_credential_detector/src/main/java/com/google/tsunami/plugins/detectors/credentials/genericweakcredentialdetector/testers/grafana/GrafanaCredentialTester.java

```diff
@@ -0,0 +1,181 @@
+import com.google.tsunami.proto.NetworkService;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+
+import javax.inject.Inject;
+import java.io.IOException;
+import java.util.Base64;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.tsunami.common.net.http.HttpRequest.get;
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+/** Credential tester specifically for grafana. */
+public final class GrafanaCredentialTester extends CredentialTester {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+  private final HttpClient httpClient;
+
+  private static final String GRAFANA_SERVICE =  "grafana";
+  private static final String GRAFANA_PAGE_TITLE =  "grafana";
+
+  @Inject
+  GrafanaCredentialTester(HttpClient httpClient) {
+    this.httpClient = checkNotNull(httpClient);
+  }
+
+  @Override
+  public String name() {
+    return "GrafanaCredentialTester";
+  }
+
+  @Override
+  public String description() {
+    return "Grafana credential tester.";
+  }
+
+  /**
+   * Determines if this tester can accept the {@link NetworkService} based on the name of the service or a custom fingerprint.
+   * The fingerprint is necessary since nmap doesn't recognize a grafana instance correctly.
+   * @param networkService the network service passed by tsunami
+   * @return true if a grafana instance is recognized
+   */
+  @Override
+  public boolean canAccept(NetworkService networkService) {
+
+    String uriAuthority = NetworkEndpointUtils.toUriAuthority(networkService.getNetworkEndpoint());
+
+    boolean canAcceptByNmapReport = NetworkServiceUtils.getWebServiceName(networkService).equals(GRAFANA_SERVICE);
+    boolean canAcceptByCustomFingerprint = false;
+
+    var url = String.format("http://%s/", uriAuthority);
+    try {
+      logger.atInfo().log("probing Grafana home - custom fingerprint phase");
+
+      HttpResponse response = httpClient.send(get(url).withEmptyHeaders().build());
+      canAcceptByCustomFingerprint = response.status().isSuccess()
+              && response
+              .bodyString()
+              .map(GrafanaCredentialTester::bodyContainsGrafanaElements)
+              .orElse(false);
+    } catch (IOException e) {
+      logger.atWarning().withCause(e).log("Unable to query '%s'.", url);
+      return false;
+    }
+
+    return canAcceptByNmapReport || canAcceptByCustomFingerprint;
+  }
+
+  // Checks if the response body contains elements of a grafana page - custom fingerprinting phase
+  private static boolean bodyContainsGrafanaElements(String responseBody) {
+    Document doc = Jsoup.parse(responseBody);
+    String title = doc.title();
+
+    if (title.toLowerCase().contains(GRAFANA_PAGE_TITLE)) {
+      logger.atInfo().log("Grafana instance probably found");
```

Comment:
Perhaps make the log a bit more specific "Found Grafana endpoint (GRAFANA_PAGE_TITLE string present in the title)"

Reply:
fixed